### PR TITLE
Support for CMA-ES minimizer based on libcmaes

### PR DESCRIPTION
**Pull request by beniz**
This is ROOT6 support for [CMA-ES](https://www.lri.fr/~hansen/cmaesintro.html), a state-of-the-art black box stochastic minimizer. The implementation uses [libcmaes](https://github.com/beniz/libcmaes), a novel high performance C++11 (with Python bindings) library for blackbox optimization.

The new minimizer yields better results than Minuit2 for most problems, though for an higher computation cost on average. See below for links to relevant benchmarks backing up these claims.

This implementation is on behalf of [Inria Saclay Research group TAO](http://www.inria.fr/en/teams/tao), [Laboratoire de l'Accelerateur Lineaire, group AppStat](http://appstat.lal.in2p3.fr/) and [University Paris-Sud LRI](https://www.lri.fr/index_en.php?lang=EN)

===Features===
- Seamless replacement for Minuit, Minuit2 and Fumili
- Seamless support for RooFit
- Relying on libcmaes allows for best performances known for CMA-ES + access to several flavors of the original algorithm, yielding best results with a trade off for computational cost, as needed
- Compilation support for both Autotools and CMake
- Included tutorial files, tests, and a special benchmark vs Minuit2

===Documentation===
- Main documentation for building and using the new minimizer is here: https://github.com/beniz/libcmaes/wiki/using-CMA-ES-in-CERN's-ROOT

===Performances===
- On-par with Minuit2 on low dimensional problems (< 10-D), consistently beats Minuit2 in higher dimension, leading to better fits. These claims are backed by two benchmarks and two experiments on real world data (we are still waiting from some results from usage at CERN)
- Benchmark CMA-ES vs Minuit on [BBOB](http://coco.gforge.inria.fr/doku.php?id=bbob-2013) available here: https://drive.google.com/open?id=0B3J1vWYhta9ibktXc2JLRUExUTA&authuser=0
- In-ROOT benchmark vs Minuit2 on low-dimensional problems, results available here: https://drive.google.com/open?id=0B3J1vWYhta9iTmR0T0hnN21lSGM&authuser=0
- Beats out Minuit2 up to 98% of the time on some experiments (data + code available on demand from Auger group, ask me)

I am available for help in testing and possibly integrating and maintaining this addition, just let me know.

### CODE REVIEW

**Comment by beniz** [comment]
Hi, I've added the README.md file for libcmaes support in root/math/cmaes as requested. Though I'm not certain this is the correct place for the readme.

Please let me know what is the current status on this PR. I am able to rebase with current head as needed for integration.

**Comment by beniz** [comment]
There's a new libcmaes 0.9.4 version that has been successfully tested against this PR to Root. Most importantly, it implements a workaround for a bug in clang, thus allowing libcmaes to be successfully built  with clang.

**Comment by beniz** [comment]
FYI, this PR has been updated to match ROOT's current master branch in case you are still interested in merging the new capabilities. Also, a new version of the libcmaes is coming up with fixes and improvements.

**Comment by beniz** [comment]
This PR has been updated for the last release of libcmaes (v0.9.5) and rebased to the tip of the ROOT master branch.

**Comment by vgvassilev** [comment]
@beniz, could you rebase again. We have better build infrastructure now and we could speed up the integration process.

File: cmake/modules/FindLibCmaes.cmake~ [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# - Try to find libcmaes, https://github.com/beniz/libcmaes
```
Comment (vgvassilev) [inline cmake/modules/FindLibCmaes.cmake~]: It seems this file is not needed. It is some editor backup.

File: cmake/modules/RootBuildOptions.cmake [hunk 2]
```diff
@@ -137,6 +137,7 @@ ROOT_BUILD_OPTION(krb5 ON "Kerberos5 support, requires Kerberos libs")
 ROOT_BUILD_OPTION(ldap ON "LDAP support, requires (Open)LDAP libs")
 ROOT_BUILD_OPTION(mathmore ON "Build the new libMathMore extended math library, requires GSL (vers. >= 1.8)")
 ROOT_BUILD_OPTION(memstat ${memstat_defvalue} "A memory statistics utility, helps to detect memory leaks")
+ROOT_BUILD_OPTION(libcmaes OFF "Build the libcmaes minimizer library")
```
Comment (vgvassilev) [inline cmake/modules/RootBuildOptions.cmake]: We'd like alphabetical order here.

File: cmake/modules/RootConfiguration.cmake [hunk 3]
```diff
@@ -388,6 +388,7 @@ set(buildmathmore ${value${mathmore}})
 set(buildcling ${value${cling}})
 set(buildroofit ${value${roofit}})
 set(buildminuit2 ${value${minuit2}})
+set(buildcmaes ${value${libcmaes}})
```
Comment (vgvassilev) [inline cmake/modules/RootConfiguration.cmake]: Likewise.

File: math/cmaes/README.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
+### This is support for black-box optimization with CMA-ES from within ROOT (http://root.cern.ch/drupal/).
```
Comment (vgvassilev) [inline math/cmaes/README.md]: Pleas update the root website link. We don't have `drupal` suffix anymore.

File: math/cmaes/inc/CMAESMinimizer.h [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with libcmaes.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef ROOT_CMAES_CMAESMinimizer
+#define ROOT_CMAES_CMAESMinimizer
+
+#ifndef ROOT_Math_Minimizer
```
Comment (vgvassilev) [inline math/cmaes/inc/CMAESMinimizer.h]: We don't use these include protectors anymore.

**Comment by beniz** [comment]
@vgvassilev Hi, I will but it may take some time given that this PR is two years old. I may require your help if I hit merging difficulties. cc @nikohansen @kegl

**Comment by vgvassilev** [comment]
Sure! I'd really like to split this PR in atomic sub PRs. This would make me and the other reviewers much more comfortable.

**Comment by beniz** [comment]
The PR basically includes:
- building changes for including CMA-ES
- the CMA-ES optimizer via libcmaes
- test samples

Not sure how to split it. I'll provide a rebase, and you guys may be able to move forward from there.

**Comment by lmoneta** [comment]
I don't think this PR should be split.
 It is fine as now. It needs some cleanup, there are some changes as in TPluginManger.cxx, TF1.cxx etc.. that are not needed. 

The only think is that the support for configure/make can be dropped (we do not support for new packages added in ROOT) and just keep the CMake support. 

If you give me again the instructions on how to build the cma-es library I can test this PR. 
On which platforms can the CMA-ES library be built ? 

Lorenzo

**Comment by beniz** [comment]
@lmoneta @vgvassilev Hi, I am closing this PR and I will provide another fresh one instead with all relevant commits added to the tip of current master.

**Comment by omazapa** [comment]
Great! I am very interested in do a review of this code after the new updates if @lmoneta  @vgvassilev  are agreed, CMA-ES appear to be a great minimizer.

Best,
Omar.

**Comment by beniz** [comment]
@lmoneta @vgvassilev @omazapa Please see PR #507 as a follow up. Your help is needed on some minor issues.

**Pull request closed without merging**